packages/next/build/webpack/plugins/next-trace-entrypoints-plugin.ts

```diff
@@ -0,0 +1,239 @@
+
+            compilation.entries.forEach((entry) => {
+              const name = entry.name || entry.options?.name
+
+              if (name?.startsWith('pages/') && entry.dependencies[0]) {
+                const entryMod = getModuleFromDependency(
+                  compilation,
+                  entry.dependencies[0]
+                )
+
+                if (entryMod.resource) {
+                  entryNameMap.set(entryMod.resource, name)
+                  entryModMap.set(entryMod.resource, entryMod)
+                }
+              }
+            })
+
+            // TODO: investigate allowing non-sync fs calls in node-file-trace
+            // for better performance
+            const readFile = (path: string) => {
+              const mod = depModMap.get(path) || entryModMap.get(path)
+
+              // map the transpiled source when available to avoid
+              // parse errors in node-file-trace
+              const source = mod?.originalSource()
+
+              if (source?._valueAsBuffer) {
+                return source._valueAsBuffer
+              }
+
+              try {
+                return compilation.inputFileSystem.readFileSync(path)
+              } catch (e) {
+                if (e.code === 'ENOENT' || e.code === 'EISDIR') {
+                  return null
+                }
+                throw e
+              }
+            }
+            const readlink = (path: string) => {
+              try {
+                return compilation.inputFileSystem.readlinkSync(path)
+              } catch (e) {
+                if (
+                  e.code !== 'EINVAL' &&
+                  e.code !== 'ENOENT' &&
+                  e.code !== 'UNKNOWN'
+                ) {
+                  throw e
+                }
+                return null
+              }
+            }
+            const stat = (path: string) => {
+              try {
+                return compilation.inputFileSystem.statSync(path)
+              } catch (e) {
+                if (e.code === 'ENOENT') {
+                  return null
+                }
+                throw e
+              }
+            }
+
+            const nftCache = {}
+            const entryPaths = Array.from(entryModMap.keys())
+
+            for (const entry of entryPaths) {
+              depModMap.clear()
+              const entryMod = entryModMap.get(entry)
+              const cachedTraces = entryMod.buildInfo?.cachedNextEntryTrace
+
+              // Use cached trace if available and trace version matches
+              if (
+                isWebpack5 &&
+                cachedTraces &&
+                cachedTraces.version === TRACE_OUTPUT_VERSION
+              ) {
+                this.entryTraces.set(
+                  entryNameMap.get(entry)!,
+                  cachedTraces.tracedDeps
+                )
+                continue
+              }
+
+              for (const dep of entryMod.dependencies) {
```

Comment:
I don't really understand what's going on here.

Why is the module graph followed only one level? What's about dependencies of dependencies?

Is this piece of code even needed?

Reply:
We need to be able to trace the transpiled version of any imported files e.g. `import util from '../util'`, I added iterating over all nested dependencies